lib/rex/proto/http/web_socket/amazon_ssm.rb

```diff
@@ -0,0 +1,315 @@
+      end
+      sbuf << '-'
+      [14...16].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf << '-'
+      [0...2].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf << '-'
+      [2...8].each do |idx|
+        sbuf << Rex::Text.to_hex(bbuf[idx])
+      end
+      sbuf.gsub("\\x",'')
+    end
+
+    def self.pack(sbuf)
+      parts = sbuf.split('-').map do |seg|
+        seg.chars.each_slice(2).map {|e| "\\x#{e.join}"}.join
+      end
+      [3, 4, 0, 1, 2].map do |part|
+        Rex::Text.hex_to_raw(parts[part])
+      end.join
+    end
+
+    def self.rand
+      self.unpack(Rex::Text.rand_text(16))
+    end
+  end
+  module Interface
+    module SsmChannelMethods
+      attr_accessor :rows
+      attr_accessor :cols
+
+      def _start_ssm_keepalive
+        @keepalive_thread = Rex::ThreadFactory.spawn('SsmChannel-Keepalive', false) do
+          while not closed? or @websocket.closed?
+            write ''
+            Rex::ThreadSafe.sleep(::Random.rand * 10 + 15)
+          end
+          @keepalive_thread = nil
+        end
+      end
+
+      def close
+        @keepalive_thread.kill if @keepalive_thread
+        @keepalive_thread = nil
+        super
+      end
+
+      def acknowledge_output(output_frame)
+        ack = output_frame.to_ack
+        # ack.header.sequence_number = @out_seq_num
+        @websocket.put_wsbinary(ack.to_binary_s, {mask: false})
+        # wlog("SsmChannel: acknowledge output #{output_frame.uuid}")
+        output_frame.uuid
+      end
+
+      def handle_output_data(output_frame)
+        if @ack_outputs.include?(output_frame.uuid)
+          # wlog("SsmChannel: repeat output #{output_frame.uuid}")
+        else
+          @ack_outputs << acknowledge_output(output_frame)
+          # TODO: handle Payload::* types
+          if [PayloadType::Output, PayloadType::Error].any? {|e| e == output_frame.payload_type }
+            if @filter_echo.is_a?(String) and output_frame.payload_data.strip == @filter_echo.strip
+              wlog("SsmChannel: filtering output #{@filter_echo}")
+              @filter_echo = true
+              return nil
+            else
+              return output_frame.payload_data
+            end
+          else
+            wlog("SsmChannel got unhandled output payload type: #{Payload.front_val(output_frame.payload_type)}")
+          end
+        end
+        nil
+      end
+
+      def handle_acknowledge(ack_frame)
+        # wlog("SsmChannel: got acknowledge message #{ack_frame.uuid}")
+        begin
+          seq_num = JSON.parse(ack_frame.payload_data)['AcknowledgedMessageSequenceNumber'].to_i
+          @ack_seq_num = seq_num if seq_num > @ack_seq_num
+        rescue => e
+          elog("SsmChannel failed to parse ack JSON #{ack_frame.payload_data} due to #{e}!")
+        end
+        nil
+      end
+
+      def update_term_size
+        rows, cols = ::IO.console.winsize
```

Comment:
Getting a stack trace here. It probably has something to do with me running it in RubyMine with the debugger .

```
msf6 auxiliary(cloud/aws/enum_ssm) > set GET_SSM_SESSION true
GET_SSM_SESSION => true
msf6 auxiliary(cloud/aws/enum_ssm) > run

[*] Checking us-east-2...
[+] {
  "AgentType": "amazon-ssm-agent",
  "AgentVersion": "3.1.1732.0",
  "ComputerName": "ip###.us-east-2.compute.internal",
  "InstanceId": "i-???",
  "InstanceStatus": "Active",
  "IpAddress": "###",
  "PlatformName": "Amazon Linux",
  "PlatformType": "Linux",
  "PlatformVersion": "2",
  "ResourceType": "EC2Instance"
}
[-] Auxiliary failed: NoMethodError undefined method `winsize' for nil:NilClass
[-] Call stack:
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/lib/rex/proto/http/web_socket/amazon_ssm.rb:123:in `update_term_size'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:151:in `get_ssm_socket'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:109:in `block (2 levels) in run'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:106:in `each'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:106:in `block in run'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:78:in `each'
[-]   /home/smcintyre/Repositories/metasploit-framework.pr/modules/auxiliary/cloud/aws/enum_ssm.rb:78:in `run'
[*] Auxiliary module execution completed
msf6 auxiliary(cloud/aws/enum_ssm) > 
``

Reply:
Neat! Definitely didn't think of that one. The pattern was stolen out of `lib/rex/post/meterpreter/ui/console/interactive_channel.rb` which i'm guessing will fail in the same manner. Since this is not a nominal use of the code, thinking we wrap it with an exception handler as its a "transparent background action" which can afford to fail. Thoughts?